app/models/transformation_mapping/vm_migration_validator.rb

```diff
@@ -0,0 +1,149 @@
+        else
+          invalid_list << VmMigrateStruct.new(vm.name, vm, VM_INVALID, reason)
+        end
+      else
+        vms.each { |v| conflict_list << VmMigrateStruct.new(v.name, v, VM_CONFLICT, VM_CONFLICT) }
+      end
+    end
+
+    {
+      "valid"      => valid_list,
+      "invalid"    => invalid_list,
+      "conflicted" => conflict_list
+    }
+  end
+
+  def validate_vm(vm, quick = true)
+    validate_result = vm_migration_status(vm)
+    return validate_result unless validate_result == VM_VALID
+
+    # a valid vm must find all resources in the mapping and has never been migrated
+    invalid_list = []
+
+    invalid_storages = vm.datastores - mapped_storages
+    if invalid_storages.present?
+      invalid_list << "storages: %{list}" % {:list => invalid_storages.collect(&:name).join(", ")}
+      return no_mapping_msg(invalid_list) if quick
+    end
+
+    invalid_lans = vm.lans - mapped_lans
+    if invalid_lans.present?
+      invalid_list << "lans: %{list}" % {:list => invalid_lans.collect(&:name).join(', ')}
+      return no_mapping_msg(invalid_list) if quick
+    end
+
+    invalid_list.present? ? no_mapping_msg(invalid_list) : VM_VALID
+  end
+
+  def vm_migration_status(vm)
+    return VM_INACTIVE unless vm.active?
+
+    vm_as_resources = ServiceResource.joins(:service_template).where(:resource => vm, :service_templates => {:type => 'ServiceTemplateTransformationPlan'})
+
+    # VM has not been migrated before
+    return VM_VALID if vm_as_resources.empty?
+
+    return VM_MIGRATED unless vm_as_resources.where(:status => ServiceResource::STATUS_COMPLETED).empty?
+
+    # VM failed in previous migration
+    vm_as_resources.all? { |rsc| rsc.status == ServiceResource::STATUS_FAILED } ? VM_VALID : VM_IN_OTHER_PLAN
+  end
+
+  def no_mapping_msg(list)
+    "Mapping source not found - %{list}" % {:list => list.join('. ')}
+  end
+
+  def mapped_clusters
+    @mapped_clusters ||= EmsCluster.where(:id => @mapping.transformation_mapping_items.where(:source_type => 'EmsCluster').pluck(:source_id))
+  end
+
+  def mapped_storages
+    @mapped_storages ||= Storage.where(:id => @mapping.transformation_mapping_items.where(:source_type => 'Storage').pluck(:source_id))
+  end
+
+  def mapped_lans
+    @mapped_lans ||= Lan.where(:id => @mapping.transformation_mapping_items.where(:source_type => 'Lan').pluck(:source_id))
+  end
+
+  class VmMigrateStruct < MiqHashStruct
+    def initialize(vm_name, vm, status, reason)
+      options = {"name" => vm_name, "status" => status, "reason" => reason}
+
+      if vm.present?
+        options.merge!(
+          "cluster"        => vm.ems_cluster.try(:name) || '',
+          "path"           => vm.ext_management_system ? "#{vm.ext_management_system.name}/#{vm.parent_blue_folder_path(:exclude_non_display_folders => true)}" : '',
+          "allocated_size" => vm.allocated_disk_storage,
+          "id"             => vm.id
```

Comment:
please change this to `vm.id.to_s` 

since on the frontend, we prefer strings.